bpo-47084: Clear cached representations on finalization

jkloth opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-47084](https://bugs.python.org/issue47084) -->
https://bugs.python.org/issue47084
<!-- /issue-number -->

jkloth committed:
Clear cached representations on finalization

jkloth committed:
Merge branch 'main' of https://github.com/python/cpython into issue47084

jkloth committed:
Also reset representation lengths on finalization

vstinner commented on Lib/test/test_embed.py:
```diff
@@ -1645,24 +1645,28 @@ def test_frozenmain(self):
+            ('off', 'import __hello__'),
+            ('on', 'import __hello__'),
+        )
+        for flag, stmt in tests:
```
Can you please use `with self.subTest(flag=flag, stmt=stmt):`? Otherwise, it will be more complicated to debug if a test fails. With that, you can omit `f"{xopt}, stmt='{stmt}' -> {out}")` below.

jkloth replied:
Didn't know about `TestCase.subTest()`, thanks.

vstinner commented on Objects/unicodeobject.c:
```diff
@@ -16070,6 +16070,58 @@ _PyUnicode_Fini(PyInterpreterState *interp)
+
+
+void
+_PyStaticUnicode_Dealloc(PyObject *op)
```
Would it make sense to share code between this function and unicode_dealloc()? Maybe add an unicode_dealloc_common() function?

jkloth replied:
Yes, there is the duplication of the `PyObect_Free()` calls, but not really to the level of what I would consider "in common".  I did, however, refactor out the duplication of the 1-character singletons and frozen strings deallocation into a standalone static function.

jkloth committed:
Merge branch 'main' of https://github.com/python/cpython into issue47084

jkloth committed:
Use TestCase.subTest() for per-iteration testing

jkloth committed:
Refactor dealloc into single function.

What started out as different behavior for each type ended up
being identical after edge cases were discovered.
